lms/static/scripts/frontend_apps/components/BookSelector.js

```diff
@@ -0,0 +1,156 @@
+import {
+  IconButton,
+  SvgIcon,
+  TextInput,
+  TextInputWithButton,
+  Thumbnail,
+} from '@hypothesis/frontend-shared';
+import { createElement } from 'preact';
+import { useRef, useState } from 'preact/hooks';
+
+import { useService, VitalSourceService } from '../services';
+
+/**
+ * @typedef {import('../api-types').Book} Book
+ */
+
+/**
+ * @typedef BookSelectorProps
+ * @prop {Book|null} selectedBook
+ * @prop {(b: Book|null) => void} onSelectBook - Callback invoked when user selects a book
+ */
+
+/**
+ * Component that prompts a user to enter/paste a URL to a VitalSource book.
+ *
+ * @param {BookSelectorProps} props
+ */
+export default function BookSelector({ selectedBook, onSelectBook }) {
+  const vsService = useService(VitalSourceService);
+
+  const inputRef = useRef(/** @type {HTMLInputElement|null} */ (null));
+  const [isLoading, setIsLoading] = useState(false);
+
+  const [error, setError] = useState(/** @type {string|null} */ (null));
+  const previousURL = useRef(/** @type {string|null} */ (null));
+
+  /**
+   * A naive regex matcher for a VBID in a URL-like string
+   *
+   * @param {string} url
+   * @returns {string|false}
+   */
+  const bookIDFromURL = url => {
+    const bookIdPattern = /book\/([0-9A-Z-]+)\/?/;
+    const matches = url.match(bookIdPattern);
+    return !!matches && matches[1];
+  };
+
+  /**
+   * Attempt to retrieve a book by bookID (vbid) via service
+   *
+   * @param {string} bookID
+   */
+  const fetchBook = bookID => {
+    setIsLoading(true);
+    vsService
+      .fetchBook(bookID)
+      .then(onSelectBook)
+      .catch(err => setError(err.message))
+      .finally(() => setIsLoading(false));
```

Comment:
`finally` is not technically supported by some of the oldest browsers in our baseline (Edge 17, Safari 10.1). We could polyfill this with core-js or we you could use `try/catch/finally` inside an async/await block.

We probably can upgrade our browser baseline to a level where `Promise.prototype.finally` is natively supported soon, but we need to gather some up-to-date usage stats using [these scripts](https://github.com/hypothesis/user-agent-analysis).

Reply:
Thanks for doing this legwork. It was tickling the back of my mind when I used `finally` that it might not be safe to do so. I'll follow up.

Reply:
I've converted this to `async/await` form.